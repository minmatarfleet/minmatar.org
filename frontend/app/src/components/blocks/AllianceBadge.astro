---
import { i18n } from '@helpers/i18n'
const { t } = i18n(Astro.url)

import type { AllianceBadgeProps } from '@dtypes/layout_components'
import { get_alliance_by_id } from '@helpers/api.eveonline/alliances'
import type { AllianceEvE } from '@dtypes/api.eveonline'
import { prod_error_messages } from '@helpers/env'

interface Props {
    alliance: AllianceBadgeProps;
}

const {
    alliance,
} = Astro.props

let alliance_eve:AllianceEvE | null = null
let get_eve_alliance_error
if (!alliance?.name) {
    try {
        alliance_eve = await get_alliance_by_id(alliance.id)
    } catch (error) {
        get_eve_alliance_error = (prod_error_messages() ? t('get_eve_alliance_error') : error.message)
    }
}

const alliance_name = alliance?.name ?? (alliance_eve?.name ?? get_eve_alliance_error)

import { get_alliance_logo } from '@helpers/eve_image_server';
import FixedFluid from '@components/compositions/FixedFluid.astro'
---
<FixedFluid
    width='32px'
    gap='var(--space-2xs)'
    class="[ items-center ]"
>
    <picture>
        <img loading="lazy" src={get_alliance_logo(alliance.id)} width="32" height="32" />
    </picture>
    <small>
        <span class="[ text-[var(--highlight)] ]">{alliance_name}</span>
    </small>
</FixedFluid>