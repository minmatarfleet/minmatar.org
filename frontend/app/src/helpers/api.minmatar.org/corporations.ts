import type { Corporation, CorporationType, CorporationInfo } from '@dtypes/api.minmatar.org'
import { get_error_message } from '@helpers/string'

const API_ENDPOINT =  `${import.meta.env.API_URL}/api/eveonline/corporations`

export async function get_all_corporations(corporation_type:CorporationType) {
    const headers = {
        'Content-Type': 'application/json'
    }

    const ENDPOINT = `${API_ENDPOINT}/corporations?corporation_type=${corporation_type}`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ));
        }

        return await response.json() as Corporation[];
    } catch (error) {
        throw new Error(`Error fetching corporations: ${error.message}`);
    }
}

export async function get_corporation_by_id(access_token:string, id:number) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/corporations/${id}`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ));
        }

        return await response.json() as Corporation;
    } catch (error) {
        throw new Error(`Error fetching corporation: ${error.message}`);
    }
}

export async function get_corporation_info(id:number) {
    const headers = {
        'Content-Type': 'application/json',
    }

    const ENDPOINT = `${API_ENDPOINT}/corporations/${id}/info`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ));
        }

        return await response.json() as CorporationInfo;
    } catch (error) {
        throw new Error(`Error fetching corporation: ${error.message}`);
    }
}