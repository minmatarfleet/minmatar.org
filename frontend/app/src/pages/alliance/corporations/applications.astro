---
import { i18n } from '@helpers/i18n'
const { lang, t, translatePath } = i18n(Astro.url)

import { HTTP_403_Forbidden } from '@helpers/http_responses'

import type { User } from '@dtypes/jwt'
import * as jose from 'jose'

const auth_token = Astro.cookies.has('auth_token') ? (Astro.cookies.get('auth_token')?.value as string) : null
const user = auth_token ? jose.decodeJwt(auth_token) as User : false
const is_superuser = (user ? user?.is_superuser : false)

import { get_user_permissions } from '@helpers/permissions'
const user_permissions = (user ? await get_user_permissions(user?.username) : [])

const can_view_page = is_superuser || user_permissions.includes('applications.change_evecorporationapplication')

if (!can_view_page)
    return HTTP_403_Forbidden()

import { get_applications_data } from '@helpers/pages/applications'
import type { ApplicationsData } from '@helpers/pages/applications'

let applications_data:ApplicationsData = {}
let data_fetching_error

try {
    applications_data = await get_applications_data(auth_token as string, lang)
} catch (error) {
    data_fetching_error = error.message
}

const {
    corporations_applications = [],
    corporations_unfiltered = [],
    corporations_applications_count = [],
    total_applications = 0,
    corporations_options = [],
} = applications_data

const CORPORATIONS_APPLICATIONS_PARTIAL_URL = translatePath('/partials/corporations_applications_component')

import Viewport from '@layouts/Viewport.astro';

import PageDefault from '@components/page/PageDefault.astro';
import PageTitle from '@components/page/PageTitle.astro';

import FlexInline from '@components/compositions/FlexInline.astro';
import Flexblock from '@components/compositions/Flexblock.astro';

import CorporationApplicationsList from '@components/blocks/CorporationApplicationsList.astro';
import TextBox from '@components/layout/TextBox.astro';
import Input from '@components/blocks/Input.astro';
import Select from '@components/blocks/Select.astro';
import ErrorRefetch from '@components/blocks/ErrorRefetch.astro';
import Button from '@components/blocks/Button.astro';

import MagnifierIcon from '@components/icons/buttons/MagnifierIcon.astro';

const page_title = t('corporations.applications.page_title')
---

<Viewport
    title={page_title}
    components={{
        alert_dialog: true,
        confirm_dialog: true,
    }}
>
    <PageDefault
        cover={{
            image: "/images/alliance-cover.jpg",
            image_990: "/images/alliance-cover.jpg",
            animated: false,
            scrollable: true,
            overlay: true
        }}
        x-data={`{
            global_collapse: false,
            corporations_applications: ${JSON.stringify(corporations_applications)},
            corporations_unfiltered: ${JSON.stringify(corporations_unfiltered)},
            corporations_filtered: ${JSON.stringify(corporations_unfiltered)},
            corporations_filtered_applications_count: ${JSON.stringify(corporations_applications_count)},
            total_applications_filtered: ${total_applications},
            total_applications_unfiltered: ${total_applications},
            corporations_options: ${JSON.stringify(corporations_options)},
            character_name: '',
            corporation_filter: '-1',
            filter() {
                if (this.character_name === '' && this.corporation_filter === '-1') {
                    this.corporations_filtered = this.corporations_unfiltered
                    this.corporations_filtered_applicationss_count = ${JSON.stringify(corporations_applications_count)}
                    this.total_applications_filtered = this.total_applications_unfiltered
                    return
                }

                this.corporations_filtered = {}
                this.corporations_filtered_applicationss_count = {}
                this.total_applications_filtered = 0

                this.corporations_applications.forEach( (corporation) => {
                    if (this.corporation_filter !== '-1' && parseInt(this.corporation_filter) !== corporation.corporation_id) {
                        this.corporations_filtered[corporation.corporation_id] = []
                        this.corporations_filtered_applications_count[corporation.corporation_id] = 0
                        return true
                    }

                    const applications = corporation.applications.filter( (application) => application.character_name.toLowerCase().includes(this.character_name.toLowerCase()) )
                    
                    if (applications.length > 0) {
                        this.total_applications_filtered += applications.length
                        this.corporations_filtered[corporation.corporation_id] = applications.map( (application) => application.id )
                        this.corporations_filtered_applications_count[corporation.corporation_id] = applications.length
                    }
                })
            },
        }`}
    >
        <FlexInline slot="header" justification='space-between' class="[ w-full ]">
            <PageTitle>
                {page_title}
            </PageTitle>
            <FlexInline>
                <Input type="text"
                    placeholder={t('filter')}
                    x-model="character_name"
                    x-on:keyup="filter()"
                >
                    <MagnifierIcon slot="icon" />
                </Input>
                
                <div class="[ hidden 2xl:block ]">
                    <Select
                        x-model="corporation_filter"
                        width='280px'
                        x-on:change="filter()"
                    >
                        <option selected value="-1">{t('corporations')}</option>
                        <template x-for="corporation in corporations_options">
                            <option x-bind:value="corporation.value" x-text="corporation.label" />
                        </template>
                    </Select>
                </div>
            </FlexInline>
        </FlexInline>

        <FlexInline slot="subheader">
            <Button
                href={translatePath('/alliance/corporations/applications/records')}
                size='sm'
            >
                {t('view_records')}
            </Button>
        </FlexInline>

        <Flexblock gap="var(--space-xl)">
            <TextBox>
                <p>{t('corporations.applications.leading_text')}</p>
            </TextBox>

            {data_fetching_error ?
                <ErrorRefetch
                    args={{
                        partial: CORPORATIONS_APPLICATIONS_PARTIAL_URL,
                        message: data_fetching_error,
                        delay: 0,
                    }}
                />
                :
                <CorporationApplicationsList corporations={corporations_applications} />
            }
        </Flexblock>
    </PageDefault>
</Viewport>