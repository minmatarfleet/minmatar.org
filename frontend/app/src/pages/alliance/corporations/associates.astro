---
import { i18n } from '@helpers/i18n'
const { lang, t, translatePath } = i18n(Astro.url)

import type { User } from '@dtypes/jwt'
import * as jose from 'jose'

const auth_token = Astro.cookies.has('auth_token') ? (Astro.cookies.get('auth_token')?.value as string) : null
const user = auth_token ? jose.decodeJwt(auth_token) as User : null
const user_id = user?.user_id ?? null

import { get_associates_data } from '@helpers/pages/associates'
import type { AssociatesData } from '@helpers/pages/associates'

let associates_data:AssociatesData = {}
let data_fetching_error:string | null = null

try {
    associates_data = await get_associates_data(auth_token, lang, user_id)
} catch (error) {
    data_fetching_error = error
}

const {
    corporations = [],
    user_corporation_id = null,
} = associates_data

const CORPORATIONS_LIST_PARTIAL_URL = `${translatePath('/partials/corporations_list_component')}?corporation_type=associate`

import Viewport from '@layouts/Viewport.astro';

import PageDefault from '@components/page/PageDefault.astro';
import PageTitle from '@components/page/PageTitle.astro';

import Flexblock from '@components/compositions/Flexblock.astro';
import FlexInline from '@components/compositions/FlexInline.astro';

import CorporationsList from '@components/blocks/CorporationsList.astro';
import ErrorRefetch from '@components/blocks/ErrorRefetch.astro';

const page_title = t('corporations.associates.page_title');
const page_description = t('corporations.associates.leading_text');
---

<Viewport
    title={page_title}
    meta_description={page_description}
    components={{
        alert_dialog: true,
        confirm_dialog: true,
    }}
>
    <PageDefault
        wide={true}
        cover={{
            image: "/images/alliance-cover.jpg",
            image_990: "/images/alliance-cover-990.jpg",
            alt: t('alliance.cover_alt'),
            animated: false,
            scrollable: true,
            overlay: true
        }}
        x-data={`{
            corporation_name: '',
            corporations_count: ${corporations.length}
        }`}
    >
        <FlexInline slot="header" justification='space-between' class="[ w-full ]">
            <FlexInline>
                <Flexblock gap="var(--space-3xs)">
                    <PageTitle>
                        {page_title}
                    </PageTitle>
                    <small x-text={`\`\${corporations_count} \${corporations_count != 1 ? '${t('corporations').toLowerCase()}' : '${t('corporation').toLowerCase()}'}\``}>
                        {corporations.length} {corporations.length != 1 ? t('corporations').toLowerCase() : t('corporation').toLowerCase()}
                    </small>
                </Flexblock>
            </FlexInline>
        </FlexInline>

        <Flexblock gap="var(--space-xl)">
            {data_fetching_error ?
                <ErrorRefetch
                    args={{
                        partial: CORPORATIONS_LIST_PARTIAL_URL,
                        message: data_fetching_error,
                        delay: 0,
                    }}
                />
                :
                <CorporationsList
                    corporations={corporations}
                    user_corporation_id={user_corporation_id ?? 0}
                    disable_applications={true}
                />
            }
        </Flexblock>
	</PageDefault>
</Viewport>

<style lang="scss">
    h3 {
        margin-block-start: var(--space-l);
    }
</style>