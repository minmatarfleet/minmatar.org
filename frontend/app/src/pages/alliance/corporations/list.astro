---
import { i18n } from '@helpers/i18n'
const { lang, t, translatePath } = i18n(Astro.url)

import type { User } from '@dtypes/jwt'
import * as jose from 'jose'

const auth_token = Astro.cookies.has('auth_token') ? (Astro.cookies.get('auth_token')?.value as string) : null
const user = auth_token ? jose.decodeJwt(auth_token) as User : null
const user_id = user?.user_id ?? null

import { get_corporations_list_data } from '@helpers/pages/corporations_list'
import type { CorporationsListData } from '@helpers/pages/corporations_list'

let corporations_list_data:CorporationsListData = {}
let data_fetching_error:string | null = null

try {
    corporations_list_data = await get_corporations_list_data(auth_token, lang, user_id)
} catch (error) {
    data_fetching_error = error
}

const {
    corporations = [],
    user_corporation_id = null,
} = corporations_list_data

const CORPORATIONS_SLIDER_PARTIAL_URL = translatePath('/partials/corporations_slider_component/')

import Viewport from '@layouts/Viewport.astro';

import PageLanding from '@components/page/PageLanding.astro';

import CorporationsSlider from '@components/blocks/CorporationsSlider.astro';
import ErrorRefetch from '@components/blocks/ErrorRefetch.astro';
import VerticalCenter from '@components/blocks/VerticalCenter.astro';

const page_title = t('corporations.list.page_title');
---

<Viewport
    title={page_title}
    meta_description={t('corporations.list.meta_description')}
    components={{
        alert_dialog: true,
        confirm_dialog: true,
    }}
>
    <PageLanding
        wide={true}
        fullscreen={true}
        cover={{
            image: "/images/corporations-cover.jpg",
            image_990: "/images/corporations-cover-990.jpg",
            alt: t('corporations.cover_alt'),
            animated: false,
            scrollable: true,
            overlay: true
        }}
        x-data={`{
            corporation_name: '',
        }`}
    >
        <VerticalCenter class="[ min-h-[90vh] ]" gap="var(--space-xl)">
            {data_fetching_error ?
                <ErrorRefetch
                    args={{
                        partial: CORPORATIONS_SLIDER_PARTIAL_URL,
                        message: data_fetching_error,
                        delay: 0,
                    }}
                />
                :
                <CorporationsSlider
                    corporations={corporations}
                    user_corporation_id={user_corporation_id ?? 0}
                />
            }
        </VerticalCenter>
	</PageLanding>
</Viewport>

<style lang="scss">
    h3 {
        margin-block-start: var(--space-l);
    }
</style>