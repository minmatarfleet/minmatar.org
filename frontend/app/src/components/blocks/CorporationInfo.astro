---
import { i18n } from '@helpers/i18n'
const { t, translatePath } = i18n(Astro.url)

interface Props {
    corporation_id:   number;
}

const {
    corporation_id,
} = Astro.props


import { get_corporation_logo } from '@helpers/eve_image_server';
import { prod_error_messages } from '@helpers/env'
import type { CorporationInfo } from '@dtypes/api.minmatar.org'
import { get_corporation_info } from '@helpers/api.minmatar.org/corporations'
import { parse_eve_html } from '@helpers/eve'
import { is_html } from '@helpers/string'

let corporation:CorporationInfo | null = null
let get_corporation_info_error:string | false = false

try {
    corporation = await get_corporation_info(corporation_id)
} catch (error) {
    get_corporation_info_error = (prod_error_messages() ? t('get_corporation_info_error') : error.message)
}

const CORPORATION_MODAL_PARTIAL_URL = translatePath('/partials/corporation_modal_component')
const delay = parseInt(Astro.url.searchParams.get('delay') ?? '0')

import { get_zkillboard_corporation_link } from '@helpers/eve'

import Wrapper from '@components/compositions/Wrapper.astro'
import Flexblock from '@components/compositions/Flexblock.astro'
import FixedFluid from '@components/compositions/FixedFluid.astro'

import AllianceBadge from '@components/blocks/AllianceBadge.astro'
import VerticalCenter from '@components/blocks/VerticalCenter.astro'
import ComponentBlock from '@components/blocks/ComponentBlock.astro'
import SectionH4 from '@components/blocks/SectionH4.astro';
import ErrorRefetch from '@components/blocks/ErrorRefetch.astro';
import Markdown from '@components/blocks/Markdown.astro';

import AmarrLogo from '@components/logos/AmarrLogo.astro';
import MinmatarLogo from '@components/logos/MinmatarLogo.astro';
import CaldariLogo from '@components/logos/CaldariLogo.astro';
import GallenteLogo from '@components/logos/GallenteLogo.astro'

import ZkillboardEvEIcon from '@components/icons/ZkillboardEvEIcon.astro';

const timezones = corporation?.timezones?.length ?? 0 > 0 ? corporation?.timezones[0].split('\n') : null
---

<Wrapper
    id="corporation-info"
    class="corporation-info overflow-y-auto"
    padding_inline='var(--space-s)'
    padding_block='var(--space-s)'
    x-data={`{
        tippy_options: {
            delay: [ 100, 200 ],
            arrow: true,
        },
        init_js() {
            tippy('[data-tippy-content]', this.tippy_options)
        }
    }`}
    x-init="init_js()"
>
    {get_corporation_info_error ?
        <ErrorRefetch
            args={{
                partial: `${CORPORATION_MODAL_PARTIAL_URL}?corporation_id=${corporation_id}`,
                message: get_corporation_info_error,
                delay: delay,
                target: '#corporation-info',
            }}
        />
        :        
        <Flexblock gap='var(--space-s)'>
            <FixedFluid
                width='96px'
                class="[ items-center ]"
            >
                <picture>
                    <img loading="lazy" width="96" height="96" src={get_corporation_logo(corporation_id, 128)} />
                </picture>
                <div class="[ fluid-fixed-l ]" class="[ w-full ]" style='--fluid-fixed--width: 32px'>
                    <VerticalCenter>
                        <Flexblock gap='var(--space-3xs)' class="[ w-full ]">
                            <Flexblock gap='0' class="[ w-full ]">
                                <span class="[ text-[var(--highlight)] ]">{corporation?.corporation_name}</span>
                                {corporation?.alliance_id &&
                                    <AllianceBadge alliance={{
                                        id: corporation.alliance_id
                                    }} />
                                }
                            </Flexblock>
                            <a
                                data-tippy-content={t('zkillboard')}
                                class="[ w-[32px] neocom-button neocom-button-focus ]"
                                href={get_zkillboard_corporation_link(corporation_id)}
                                target="_blank"
                                rel="nofollow noopener noreferrer"
                            >
                                <ZkillboardEvEIcon />
                            </a>
                        </Flexblock>
                    </VerticalCenter>
                    <VerticalCenter>
                        <picture>
                            {corporation?.faction_name === 'Amarr Empire' &&
                                <AmarrLogo />
                            }
                            {corporation?.faction_name === 'Minmatar Republic' &&
                                <MinmatarLogo />
                            }
                            {corporation?.faction_name === 'Caldari State' &&
                                <CaldariLogo />
                            }
                            {corporation?.faction_name === 'Gallente Federation' &&
                                <GallenteLogo />
                            }
                        </picture>
                    </VerticalCenter>
                </div>
            </FixedFluid>
            
            {corporation?.biography !== '' &&
                <ComponentBlock>
                    <SectionH4 heading_text={t('description')}>
                        <Markdown markdown={corporation?.biography as string} />
                    </SectionH4>
                </ComponentBlock>
            }
            {(corporation?.timezones.length ?? 0) > 0 &&
                <ComponentBlock>
                    <SectionH4 heading_text={t('timezones')}>
                        <Markdown markdown={`- ${timezones?.join('- ')}` as string} />
                    </SectionH4>
                </ComponentBlock>
            }
            {(corporation?.requirements.length ?? 0) > 0 &&
                <ComponentBlock>
                    <SectionH4 heading_text={t('requirements')}>
                        <Markdown markdown={corporation?.requirements.join('') as string} />
                    </SectionH4>
                </ComponentBlock>
            }
        </Flexblock>
    }
</Wrapper>

<style lang="scss">
    .corporation-info {
        width: 100%;
        max-width: 60rem;
        max-height: min(750px, 88vh);
        height: 100%;
        position: relative;
        background-position: top left;
        background-repeat: no-repeat;
        background-color: var(--background);

        > * {
            position: relative;
            z-index: 1;
        }
    }
</style>