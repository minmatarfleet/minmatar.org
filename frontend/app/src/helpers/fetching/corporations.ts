import { useTranslations } from '@i18n/utils';
import { get_user_character } from '@helpers/fetching/characters'
import type { EveCharacterProfile } from '@dtypes/api.minmatar.org'

const t = useTranslations('en');

import type { Corporation, CorporationApplication, CorporationType, CharacterCorp } from '@dtypes/api.minmatar.org'
import type { CorporationObject, CorporationStatusType, CorporationMembers, CharacterKind, CharacterBasic } from '@dtypes/layout_components'
import { get_all_corporations, get_corporation_by_id } from '@helpers/api.minmatar.org/corporations'
import { get_corporation_applications } from '@helpers/api.minmatar.org/applications'

export async function get_corporations_list_auth(access_token:string, user_id: number, corporation_type:CorporationType) {
    let api_corporations:Corporation[] = []
    let corporations:CorporationObject[] = []

    api_corporations = await get_all_corporations(corporation_type)

    corporations = await Promise.all(api_corporations.map(async (i) => add_status_to_corporation(access_token, i, user_id) ));
    
    return corporations
}

export async function get_corporations_list(corporation_type:CorporationType) {
    let api_corporations:Corporation[] = []
    let corporations:CorporationObject[] = []

    api_corporations = await get_all_corporations(corporation_type)

    corporations = api_corporations.map( (i):CorporationObject => {
        return {
            alliance_id: i.alliance_id,
            alliance_name: i.alliance_name,
            corporation_id: i.corporation_id,
            corporation_name: i.corporation_name,
            corporation_type: i.type,
            active: i.active,
            biography: i.biography,
            introduction: i.introduction,
            requirements: i.requirements,
            timezones: i.timezones,
            status: 'unauth'
        }
    } )
    
    return corporations
}

export async function get_corporation_list_by_id_auth(access_token:string, corporation_id:number, user_id: number) {
    let api_corporation:Corporation

    api_corporation = await get_corporation_by_id(access_token, corporation_id)

    return await add_status_to_corporation(access_token, api_corporation, user_id)
}

const add_status_to_corporation = async (access_token:string, api_corporation:Corporation, user_id:number) => {
    let corporation_applications:CorporationApplication[]

    const corporation:CorporationObject = {
        corporation_id: api_corporation.corporation_id,
        corporation_name: api_corporation.corporation_name,
        alliance_id: api_corporation.alliance_id,
        alliance_name: api_corporation.alliance_name,
        corporation_type: api_corporation.type,
        active: api_corporation.active,
        status: 'available',
        biography: api_corporation.biography,
        introduction: api_corporation.introduction,
        requirements: api_corporation.requirements,
        timezones: api_corporation.timezones,
    }

    try {
        corporation_applications = await get_corporation_applications(access_token, api_corporation.corporation_id)
    } catch (error) {
        corporation.status = 'error'
        return corporation
    }
    
    const user_application = user_id ? corporation_applications.find( (application) => application.user_id == user_id ) : undefined

    if (user_application !== undefined)
        corporation.status = user_application.status as CorporationStatusType

    return corporation
}

export async function get_user_corporation_id(user_id:number) {
    let user_character:EveCharacterProfile
    user_character = (user_id ? await get_user_character(user_id) : null)
    return !user_character ? null : (user_character?.corporation_id ?? null)
}

export async function get_all_corporations_members(access_token:string) {
    let api_corporations:Corporation[]
    let corporation_members:CorporationMembers[]

    api_corporations = await get_all_corporations('alliance')

    corporation_members = (await Promise.all(api_corporations.map(async (corporation) => 
        await get_all_corporation_members(access_token, corporation.corporation_id)
    )))

    return corporation_members
}

export async function get_all_corporation_members(access_token:string, corporation_id:number) {
    let api_corporation:Corporation
    let corporation_members:CorporationMembers

    api_corporation = await get_corporation_by_id(access_token, corporation_id)

    corporation_members = {
        corporation_id: api_corporation.corporation_id,
        corporation_name: api_corporation.corporation_name,
        active: api_corporation.active,
        type: api_corporation.type,
        members: []
    }

    corporation_members.members = (await Promise.all(api_corporation.members.map(async (api_member) => {
        const is_alt = (api_member.primary_character_id !== null)

        let member:CharacterKind = {
            character_id: api_member.character_id,
            character_name: api_member.character_name,
            registered: api_member.registered,
            exempt: api_member.exempt,
            is_main: !is_alt
        }
        
        if (is_alt) {
            member.main_character = {
                character_id: api_member.primary_character_id,
                character_name: api_member.primary_character_name,
            }
        }

        return member
    })))

    return corporation_members
}

export async function get_all_alliance_members(access_token:string) {
    let members:CharacterBasic[] = []    
    let api_corporations:Corporation[]
    let corporation_members:CorporationMembers[]

    api_corporations = await get_all_corporations('alliance')

    corporation_members = (await Promise.all(api_corporations.map(async (corporation) => 
        await get_all_corporation_members(access_token, corporation.corporation_id)
    )))

    corporation_members.map( (corporation) =>
        corporation.members.map( (member) => {
            members.push({
                character_id: member.character_id,
                character_name: member.character_name,
                corporation: {
                    id: corporation.corporation_id,
                    name: corporation.corporation_name,
                }
            })
        })
    )

    return members
}