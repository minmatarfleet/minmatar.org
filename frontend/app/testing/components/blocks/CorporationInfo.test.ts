import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test, vi } from "vitest";
import CorporationInfo from "@components/blocks/CorporationInfo.astro";

import { get_corporation_info } from "@helpers/api.minmatar.org/corporations";
import AllianceBadge from "@components/blocks/AllianceBadge.astro";
vi.mock("@helpers/api.minmatar.org/corporations");
vi.mock("@components/blocks/AllianceBadge.astro");

test("CorporationInfo defaults", async () => {
  vi.mocked(get_corporation_info).mockResolvedValue({
    corporation_id: 454,
    corporation_name: "chicken attack",
    alliance_id: 34523,
    active: true,
    alliance_name: "coop",
    faction_id: 1,
    faction_name: "minmatar",
    requirements: ["chicken"],
    biography: "",
    introduction: "",
    timezones: [],
    type: "alliance",
  });
  vi.mocked(AllianceBadge).mockReturnValue("AllianceBadge_component");

  const container = await AstroContainer.create();
  const result = await container.renderToString(CorporationInfo, {
    props: {
      corporation: {
        corporation_id: 2345,
      },
    },
    slots: {
      default: "slotted chick",
    },
  });

  expect(result).toMatchSnapshot();
});
