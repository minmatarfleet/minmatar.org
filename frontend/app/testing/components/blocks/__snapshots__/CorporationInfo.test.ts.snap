// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CorporationInfo defaults 1`] = `
"<div class="wrapper corporation-info overflow-y-auto" id="corporation-info" x-data="{
        tippy_options: {
            delay: [ 100, 200 ],
            arrow: true,
        },
        init_js() {
            tippy('[data-tippy-content]', this.tippy_options)
        }
    }" x-init="init_js()" data-astro-cid-76tefetw="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-s);--wrapper_padding_inline_end: var(--space-s);--wrapper_padding_block: var(--space-s);--wrapper_max_width: 80rem;"> <div class="flex-block gap---space-s" data-astro-cid-76tefetw="true" data-mod style="--flex_block_gap: var(--space-s);">  <div class="fixed-fluid [ items-center ]" data-astro-cid-76tefetw="true" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 96px;--fixed_fluid_breakpoint: 50%;">  <picture data-astro-cid-76tefetw> <img loading="lazy" width="96" height="96" src="https://images.evetech.net/corporations/undefined/logo?size=128" data-astro-cid-76tefetw> </picture> <div class="[ fluid-fixed-l ]" class="[ w-full ]" style="--fluid-fixed--width: 32px" data-astro-cid-76tefetw> <div class="!flex !items-center" data-astro-cid-76tefetw="true">  <div class="flex-block [ w-full ] gap---space-3xs" data-astro-cid-76tefetw="true" data-mod style="--flex_block_gap: var(--space-3xs);">  <div class="flex-block [ w-full ] gap-0" data-astro-cid-76tefetw="true" data-mod style="--flex_block_gap: 0;">  <span class="[ text-[var(--highlight)] ]" data-astro-cid-76tefetw>chicken attack</span> AllianceBadge_component </div>  <a data-tippy-content="Zkillboard" class="[ w-[32px] neocom-button neocom-button-focus ]" href="https://zkillboard.com/corporation/undefined/" target="_blank" rel="nofollow noopener noreferrer" data-astro-cid-76tefetw> <img loading="lazy" class="[ max-w-[100%] h-auto ]" src="/images/icons/48px-Combatlog.png" width="48" height="48" alt="EVE Online Combat Log Icon"> </a>  </div>   </div> <div class="!flex !items-center" data-astro-cid-76tefetw="true">  <picture data-astro-cid-76tefetw>     </picture>  </div> </div>  </div>   <div class="wrapper [ card component ] light-transparency" data-astro-cid-76tefetw="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-l-xl);--wrapper_padding_inline_end: var(--space-l-xl);--wrapper_padding_block: var(--component-padding-block);--wrapper_max_width: 100%;">   <div class="flex-block gap---space-xs" data-astro-cid-76tefetw="true" data-mod style="--flex_block_gap: var(--space-xs);">  <h4>Requirements</h4>  <div data-astro-cid-76tefetw><p>chicken</p>
</div>   </div>    </div>   </div>  </div>   "
`;
